### ISSUE TITLE
Keyboard trigger key press multiple times for Modifier keys (Shift, Control, Alt, Wooting key)

### ISSUE THREAD

**Issue by yeradis** [WootingKb/wootility-issues#19]
**TL;DR: is a problem with the nKRO enabled, when disabled the issue dissapeared**

**Describe the bug**
When pressing the Shift key or Ctrl key or Wooting Key (ui?) or Alt key the system is receiving the key press event as it was pressed twice

**To Reproduce**
Dont know how to reproduce it, this is happening even after keyboard restore
i have observed that this is happening with the keys that have a pair in the keyboard

normal behaviour:
![image](https://user-images.githubusercontent.com/666944/76414828-6215dc00-6398-11ea-8fbc-ff26057f8df8.png)

**Screenshots / Video**
key pressed and released without holding it down just one time without repetition on my side

Ctrl key:
![image](https://user-images.githubusercontent.com/666944/76414407-8ae9a180-6397-11ea-8bcd-b45cd2da77b2.png)

Shift key:

![image](https://user-images.githubusercontent.com/666944/76414496-bb314000-6397-11ea-8e3d-9e4456fd10f5.png)

Wooting key:

![image](https://user-images.githubusercontent.com/666944/76414552-d439f100-6397-11ea-965e-224ebd592497.png)

Alt:

![image](https://user-images.githubusercontent.com/666944/76414661-106d5180-6398-11ea-9b17-7d6c90a4d82d.png)

**Desktop (please complete the following information):**
OS: MacOs Catalina v10.15.3
MacBook Pro (Retina, 15-inch, Mid 2015)

**Wootility and Firmware version**
firmware version: 1.27.3
Wootility version: 3.4.5

**Keyboard Type**
Wooting one + US ANSI

**Comment by yeradis** [comment]
Just more information
If i enable Tachyon Mode, now appears as pressed three times and released just one

![image](https://user-images.githubusercontent.com/666944/76517435-9a332280-645d-11ea-9f33-ba88b5e36bb4.png)

**Comment by Rocky04** [comment]
**It seems that the problem is the nKRO keyboard interface.**

After disabling nKRO in the Wootility the problem is not present anymore for him.

**Comment by Rocky04** [comment]
Ok, so another user has problems too related to this bug in a similiar way. For him the space key wasn't working. After he disabled the nKRO keyboard interface it was working.

His system:
System Version:    macOS 10.14.5 (18F132)
Kernel Version:     Darwin 18.6.0
Model Name:       MacBook Pro
Model Identifier:  MacBookPro10,1
Processor Name:  Intel Core i7
Processor Speed: 2.6 GHz

**Comment by yeradis** [comment]
The workaround of disabling the nKRO doesnt fix the issue, also it gets worse on

Firmware version: 1.28.1
Wootility version: 3.5.0-alpha

Specifically:

nKRO disabled = keys repeated twice
nKRO enabled = keys repeated three times

**Comment by Rocky04** [comment]
You should downgrade and test it on the current Stable version again, I bet then it works again.

Because on the current Alpha only the media key interface is changed, it can be that this causes the all the trouble. So if this is possible you should disable it, it's the 4th interface on the Wooting with endpoint 0x87 and the usage page 12 and usage 1.

**Comment by PastaJ36** [comment]
Can confirm its the same on my mac, will dig into the firmware.

**Comment by Rocky04** [comment]
After he performed the downgrade to the current Stable the workaround was working again.

Because the only difference is the changed media key interface the problem is that or in combination with the nKRO beyboard interface.

**Comment by PastaJ36** [comment]
Alright, finally got around to fully digging into this issue. Short story, I was able to fix this issue for Mac and will add it to the next alpha / beta firmware.

## Long story:

The NKRO disable had a bug in the alpha version. Because of a merge conflict the setting wasn't properly set. That explains the difference between the alpha and the stable version.

About this bug: I found out that the double modifier bug has actually already been in the firmware for a very long time. To give some context: for keyboard output we use two interfaces:
1. The NKRO interface that will normally be used.
2. A "hidden" standard 6KRO interface that is meant to be used for bootloader compatibility.

The hidden interface set its report **count** to 0 in the HID descriptor. The OS should read this and then ignore any packages coming from the interface, since its report count is 0. The bootloader does not even parse the report and thus should just take up the keys sent through this interface.

The HID descriptor has two parts:
1. A part for modifier keys.
2. A part for all the rest.

It seems like Mac just ignores this report count field for the modifier part of the 6KRO descriptor. What is very weird is that it does works correctly for the 2nd part, because it does ignore the normal keys. I'm not sure why this is.

I was able to "fix" this issue by also setting the report **size** field to 0. Mac does properly parse that field and ignores both the modifiers and the rest of the keys now from the 6KRO interface. 

@Rocky04 about the spacebar thing, I suspect it's a different issue. We should open another issue if it happens again. I wasn't able to reproduce that one and it doesn't seem related to the cause of this issue.

**Comment by Rocky04** [comment]
@PastaJ36 Windows behave the same. There are always two indication messages sent to the keyboard. 

Even this is not the underlaying cause of the problem with the race condition I bet it's the reason how it's triggered because it's much easier to spam the keyboard in a way it can't handle the input messages in time anymore.
https://github.com/WootingKb/wootility-issues/issues/21

**Comment by PastaJ36** [comment]
"Windows behave the same. There are always two indication messages sent to the keyboard too." With indication messages do you mean the LED messages or that on windows there are also always two modifier keys sent?

I think we're mixing two things here. The status indicators don't seem to be related. When i talked about changing the report size fields I meant purely for the modifier and normal part of the HID descriptor. 

So your theory is that because the LED usage page is still defined, it might confuse that byte with the modifier byte? I guess it could be, but if that were the case this fix wouldn't work. Because now I only set the modifier and normal part of the HID descriptor to 0 and there's no double modifiers anymore. I don't think it's related.

I think it's Mac not properly parsing the **report count** part of the report.

That said your description did give me another idea related to the LED issue. I'll comment there to keep it separate.

**Comment by Rocky04** [comment]
No I mean that Win sent a LED message to both keyboard interfaces, to the boot interface and to the nKRO interface because the message length is still 1 byte and so valid. From my testing Win handles the length values correctly and that's the reason why this bug isn't present.

I think it's related in a way that this bug causes the other bug to occur more often (it triggers it) due to the multiple messages around the same time. So I think it's most likely related but it's not the root of the other bug.

**Comment by Rocky04** [comment]
If nKRO is disabled there are two LED status messages sent again on Win. So I don't know how the Wooting now handles that again.

**Comment by Rocky04** [comment]
Maybe it would be good to set the message size to zero for the outgoing nKRO interface if it's disabled to avoid that the systems is sending LED update messages over this interface. So that there is always ony one outgoing keyboard interface active, the 6KRO or nKRO interface.

**Comment by PastaJ36** [comment]
We could, but I try to limit the changes we do to the USB interfaces. One day if we'll switch USB stacks (for example when we go to ARM) it will be a little easier to swap out the USB stack.

**Comment by Rocky04** [comment]
Should be fixed in Wootility 3.5.2 and firmware 1.28.3.
